### PR TITLE
Infrastructure for ZM interface with AD

### PR DESCRIPTION
**Pull request by mfossitt**
This commit introduces the SCREAM-AD infrastructure to allow the ZM deep convection scheme to be run in the scream physics suite.

The ZM scheme included here is an adaptation of the ZM scheme that comes with the release version of E3SM.  The adaptations include:
1. A removal of certain CAM dependencies.
2. Switch to handling the calculation of the vapor saturation mixing ratio to follow the algorithm used by P3, for consistency.

### CODE REVIEW

File: components/scream/src/physics/zm/atmosphere_deepconvection.hpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+#include "ekat/scream_parameter_list.hpp"
+
+#include <string>
+
+namespace scream
+{
+
+/*
+ * The class responsible to handle the atmosphere microphysics
+ *
```
Comment (PeterCaldwell) [inline components/scream/src/physics/zm/atmosphere_deepconvection.hpp:14]: instead of 'microphysics', the line above should say 'deep convection', right?

File: components/scream/tests/scream_zm/CMakeFiles/3.14.5/CMakeCCompiler.cmake [hunk 2]
```diff
@@ -0,0 +1,75 @@
+set(CMAKE_C_COMPILER "/usr/tce/bin/cc")
```
Comment (PeterCaldwell) [inline components/scream/tests/scream_zm/CMakeFiles/3.14.5/CMakeCCompiler.cmake:1]: What is the motivation for changing this file? These changes seem unrelated to getting ZM working.

Comment (bartgol) [reply components/scream/tests/scream_zm/CMakeFiles/3.14.5/CMakeCCompiler.cmake:1]: Yes, this should be removed. Besides not being portable (someone might not have `cc` in that path), it is also not safe in a CMake project to change the compiler _after_ the compilers have been inspected by cmake (either at the `project` call or `enable_language` call).

File: components/scream/src/physics/zm/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,37 @@
+set(CIMEROOT ${SCREAM_BASE_DIR}/../../cime)
+list(APPEND CMAKE_MODULE_PATH ${CIMEROOT}/src/CMake)
+
+set(GENF90 ${CIMEROOT}/src/externals/genf90/genf90.pl)
+set(ENABLE_GENF90 True)
+include(genf90_utils)
+include(Sourcelist_utils)
```
Comment (bartgol) [inline components/scream/src/physics/zm/CMakeLists.txt:7]: I don't see you using genf90 on any file here. Perhaps you needed it at the beginning, then as you modified zm your dependency on `*.f90.in` files went away? If so, let's remove this stuff.

File: components/scream/src/physics/zm/atmosphere_deepconvection.cpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+void ZMDeepConvection::initialize (const util::TimeStamp& t0)
+{
+  m_current_ts = t0;
+  
+  zm_init_f90 (*m_raw_ptrs_in["limcnv_in"], m_raw_ptrs_in["no_deep_pbl_in"]);
+  
+  using strvec = std::vector<std::string>;
+  const strvec& initable = zm_inputs;
+  
+  for (const auto& name : initable) {
```
Comment (bartgol) [inline components/scream/src/physics/zm/atmosphere_deepconvection.cpp:64]: You need to think about the initialization, based on what outcome you want. As you implemented it, if any field that ZM needs is not inited, ZM will init it. One thing you may want to check is whether you're ok with ZM initing a few of the inputs, but not all. Like, if all but one fields already have an initialization, is that ok or is it a symptom of an error (like, that field was supposed to be inited alongside the others)?

You can look at P3, for an example on how to guard from that scenario.

File: components/scream/src/physics/zm/atmosphere_deepconvection.cpp [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  }
+}
+
+// =========================================================================================
+void ZMDeepConvection::run (const Real dt)
+{
+  std::vector<const Real*> in;
+  std::vector<Real*> out;
+
+  // Copy inputs to host. Copy also outputs, cause we might "update" them, rather than overwrite them.
```
Comment (bartgol) [inline components/scream/src/physics/zm/atmosphere_deepconvection.cpp:80]: I don't know how much work is in ZM, if it's "a ton", then it might mask the deep copy costs. Still, just in case: are all the outputs "updated" rather than "overwritten"? If so, that's ok. Otherwise, you might consider copying to host only the outputs that are updated.

File: components/scream/src/physics/zm/atmosphere_deepconvection.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+   for (auto& fid : m_required_fields) {
+     field_repo.register_field(fid);
+   }
+   for (auto& fid : m_computed_fields) {
+     field_repo.register_field(fid);
+   }
+ }
+
+void ZMDeepConvection::set_required_field_impl (const Field<const Real, device_type>& f) {
+  // @Meredith: Diff between add_me_as_a_customer and get_tracking().add_customer?
```
Comment (bartgol) [inline components/scream/src/physics/zm/atmosphere_deepconvection.cpp:133]: The method `add_me_as_customer(f)` is just a shortcut for

```
f.get_header_ptr()->get_tracking().add_customer(weak_from_this());
``` 
I thought it was easier to understand what was going on if someone simply read `add_me_as_customer` than the other long string. Definitely less error prone (if you use `get_header` rather than `get_header_ptr`, you get a const-protected `FieldTracking`, so you can't modify it...unless you play with `const_cast`).

File: components/scream/src/physics/zm/atmosphere_deepconvection.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+   }
+ }
+
+void ZMDeepConvection::set_required_field_impl (const Field<const Real, device_type>& f) {
+  // @Meredith: Diff between add_me_as_a_customer and get_tracking().add_customer? 
+  
+  // Store a copy of the field. We need this in order to do some tracking checks
+  // at the beginning of the run call. Other than that, there would be really
+  // no need to store a scream field here; we could simply set the view ptr
+  // in the Homme's view, and be done with it.
```
Comment (bartgol) [inline components/scream/src/physics/zm/atmosphere_deepconvection.cpp:138]: This comment seems copied from HommeDynamics. Better delete it, to not cause confusion.

File: components/scream/src/physics/zm/atmosphere_deepconvection.cpp [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    Kokkos::deep_copy(m_zm_host_views_in.at(it.first),it.second.get_view());
+  }
+  for (auto& it : m_zm_fields_out) {
+    Kokkos::deep_copy(m_zm_host_views_out.at(it.first),it.second.get_view());
+  }
+
+  Real** temp = &m_raw_ptrs_out["fracis"];
+  Real*** fracis = &temp;
+ 
+  zm_main_f90(*m_raw_ptrs_out["lchnk"], *m_raw_ptrs_out["ncol"], m_raw_ptrs_out["t"],
```
Comment (bartgol) [inline components/scream/src/physics/zm/atmosphere_deepconvection.cpp:91]: It looks like you get all fields from `m_raw_ptrs_out`. I'm guessing some (at least one) should be stored in `m_raw_ptrs_in`?

File: components/scream/src/physics/zm/scream_zm_interface.hpp [hunk 9]
```diff
@@ -0,0 +1,36 @@
+#ifndef SCREAM_ZM_INTERFACE_HPP
+#define SCREAM_ZM_INTERFACE_HPP
+
+#include "ekat/scream_assert.hpp"
```
Comment (bartgol) [inline components/scream/src/physics/zm/scream_zm_interface.hpp:4]: These files changed name in a recent PR merge where we updated the version of the ekat submodule. You will have to merge master in your branch...and cross your fingers.

Jokes aside, if you get too many conflicts and need help, or are note sure on how to resolve some of them, let me know.

File: components/scream/src/physics/zm/zm_inputs_initializer.cpp [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    "Error! ZMInputsInitializer is expected to init _______________.\n"
+    "Only " + std::to_string(count) + " of those have been found.\n"
+    "Please, check the atmosphere processes you are using,"
+    "and make sure they agree on who's initializing each field.\n");
+
+
+
+  for (int i = 0; i < zm_inputs.size(); i++){
+    //Get and store device view using input name
+    Kokkos::View<Real*, Kokkos::LayoutRight, HostDevice> d_v = m_fields.at(zm_inputs[i]).get_view();
```
Comment (bartgol) [inline components/scream/src/physics/zm/zm_inputs_initializer.cpp:226]: Using the keyword `auto` might make this line and the next a bit less "heavy". Like:

```
auto d_v = ...
auto h_m = Kokkos::create_mirror_view(d_v);
```
Also, but this is somewhat a personal taste subject, it might maker the code easier if you used a pattern like "blah_d"/"blah_h" for device/host views (or even "blah_dev"/"blah_host"), or "d_blah"/"h_blah" if you like to prepend the memory location.

File: components/scream/tests/scream_zm/CMakeFiles/3.14.5/CMakeCCompiler.cmake (hunk 2 above)
Comment (bartgol) [reply components/scream/tests/scream_zm/CMakeFiles/3.14.5/CMakeCCompiler.cmake:1]: I now see that there's a whole CMakeFiles folder. This looks like something you find in a build tree. I'm assuming you build (or built at least once) the code in the src tree, and ended up with these leftovers accidentally commited?

File: components/scream/tests/scream_zm/input.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+Debug:
+  Atmosphere DAG Verbosity Level: 5
+
+Atmosphere Processes:
+  Number of Entries: 1
+
+  Process 0:
+    Process Name: ZM
+    Grid: Physics
+    Can Initialize All Inputs: true
```
Comment (bartgol) [inline components/scream/tests/scream_zm/input.yaml:12]: You are not currently using this. But see my comment above on checking P3 approach on input initialization. Might be that you want to keep this option, and start using it, like P3 does.

File: components/scream/tests/scream_zm/CMakeFiles/3.14.5/CMakeCCompiler.cmake (hunk 2 above)
Comment (AaronDonahue) [reply components/scream/tests/scream_zm/CMakeFiles/3.14.5/CMakeCCompiler.cmake:1]: Yes, this probably comes from building and testing.  Similar to what you get when you run `test-all-scream` and forget to delete the `ctest` directory.

Comment (bartgol) [reply components/scream/tests/scream_zm/CMakeFiles/3.14.5/CMakeCCompiler.cmake:1]: Ah, btw, to avoid seeing unstaged changes like the ctest directory, you can modify the `.git/info/exclude` file in your repo, and add that directory to the list. That file works like `.gitignore`, but it is _not part of the repo_, so it does not get staged/committed. It kind of works as a "local" `.gitignore`.

File: .gitmodules [hunk 12]
```diff
@@ -34,10 +34,6 @@
 	path = components/cam/src/physics/cosp2/external
 	url = git@github.com:CFMIP/COSPv2.0.git
 	branch = CESM_v2.1.4
-[submodule "cime"]
```
[the remark this answers is not in the document]
Comment (PeterCaldwell) [reply .gitmodules:37]: It looks like submodule cime is missing completely. You also need to git submodule update following https://acme-climate.atlassian.net/wiki/spaces/ED/pages/1868455/Development+Getting+Started+Guide (maybe you already did, just mentioning in case). I don't think cime matters for what you're doing yet since it is so far only used when running with the old driver... but don't remove it's use for other folks!

File: components/scream/src/physics/zm/atmosphere_deepconvection.cpp [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+
+void ZMDeepConvection::set_grids(const std::shared_ptr<const GridsManager> grids_manager)
+{
+  using namespace std;
+
+  using namespace units;
+  auto Q = kg/kg;
+  auto nondim = m/m;
```
Comment (PeterCaldwell) [inline components/scream/src/physics/zm/atmosphere_deepconvection.cpp:23]: this must be C++ magic I don't understand: how can Q=kg/kg be a valid variable assignment? I would think you'd need quotes around kg/kg and m/m.

Comment (AaronDonahue) [reply components/scream/src/physics/zm/atmosphere_deepconvection.cpp:23]: This is totally C++ magic.  It has to do with the namespace units that is being used and it is something that Luca put in to deal with the mass mixing ratio units issue where the units are kg_X/kg_air, which is technically kg/kg.

Comment (bartgol) [reply components/scream/src/physics/zm/atmosphere_deepconvection.cpp:23]: `auto Q=kg/kg;` is a normal assignment, just like `auto x = 1.0 / 3.0;`. Right after the assignment, however, Q is a nondimensional unit (in fact, Q and nondim are the same). The following line is what causes Q to print a user defined string when outputed to screen (here "kg/kg"), while nondim would show "[1]" (or "1", I don't recall).

The `set_string` method accepts anything. One could do `Q.set_string("Luca is the best");`. There's no check on what string is passed. Since this is only used at "cout" time, it's not a big deal though if two places use two different strings (like "kg/kg" and "kg  / kg ").

Comment (PeterCaldwell) [reply components/scream/src/physics/zm/atmosphere_deepconvection.cpp:23]: Ok, that makes sense. Thanks Luca!

Comment (bartgol) [reply components/scream/src/physics/zm/atmosphere_deepconvection.cpp:23]: @PeterCaldwell sorry, I put the second comment in pretty much right after the first, but I started a review (rather than post single comment) that I didn't complete, so It didn't posted it until now (when I noticed). I hope it now makes more sense.

File: components/scream/src/physics/zm/atmosphere_deepconvection.cpp (hunk 4 above)
Comment (mfossitt) [reply components/scream/src/physics/zm/atmosphere_deepconvection.cpp:64]: Good call. Working on implementing similar checks as P3 has.

File: components/scream/src/physics/zm/atmosphere_deepconvection.cpp (hunk 8 above)
Comment (mfossitt) [reply components/scream/src/physics/zm/atmosphere_deepconvection.cpp:91]: Yes, at some point they should be. I defaulted to m_raw_ptrs_out because I was not sure which vars should be inputs/outputs as I don't know the functionality of the fortran well enough. This process will probably require conferring with someone else on the team.

File: components/scream/src/physics/zm/zm_inputs_initializer.cpp (hunk 10 above)
Comment (mfossitt) [reply components/scream/src/physics/zm/zm_inputs_initializer.cpp:226]: I actually changed the auto convention intentionally, because the auto keyword made putting values in structs and vectors difficult if not impossible. I can change the structure so that it uses auto but I think that the cleanliness of auto would mean adding more lines of more muddled code

File: components/scream/src/physics/zm/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+set(ZM_SRCS
+  ${SCREAM_BASE_DIR}/../cam/src/physics/cam/physics_utils.F90
+  ${SCREAM_BASE_DIR}/../cam/src/physics/cam/scream_abortutils.F90
+  zm_conv.F90
+  atmosphere_deepconvection.cpp
```
[the remark this answers is not in the document]
Comment (mfossitt) [reply components/scream/src/physics/zm/CMakeLists.txt:13]: For sure

**Comment by mfossitt** [comment]
@AaronDonahue TODOs:
1. Figure out which inputs are raw_ptrs_in/raw_ptrs_out
2. Figure out units for each function param

**Comment by AaronDonahue** [comment]
@PeterCaldwell should we adjust the ZM scheme here to use the now universal saturation vapor pressure calculation per PR #588?

**Comment by PeterCaldwell** [comment]
Yes, everyone should use the universal saturation vapor pressure calculation. You can make the change in this PR, or integrate this PR and make a small new PR to do the vapor pressure switch.

**Comment by AaronDonahue** [comment]
Ok great, that is what I figured.  I think we can just make an update to this PR.  I'll do an upstream merge of master into this branch and then make the wv_sat adjustment.

**Comment by AaronDonahue** [comment]
@mfossitt , I started a Confluence page to go through the variables and determine the units and in/out status.  see: https://acme-climate.atlassian.net/wiki/spaces/NGDNA/pages/1845133326/Table+of+variables+and+their+units

**Comment by AaronDonahue** [comment]
@bartgol and @PeterCaldwell , considering that this PR is only waiting on updates regarding the units and the in/out status of the field variables Meredith and I, after chatting a bit with Peter, have decided a good step forward would be to review and integrate the PR as is and post a git issue about correcting the units and getting the in/out status sorted out.  

A main motivation for this course of action is that the PR is rather old and is getting stale, the AD team is focused on more pressing deliverables so it may take some time to get to the units checking and ZM isn't a part of the main scream physics suite, so it is probably ok to keep it as is for now and correct later.

If you agree with this plan then all that is needed is the final approval of the PR.  Meredith recently did a merge from master so we anticipate no conflicts and that the AT will pass.  But we can probably wait for the AT to finish before going through the last round of reviews.

**Comment by bartgol** [comment]
@AaronDonahue @mfossitt I have a buys week, but I will try to get the review done by tomorrow. Meanwhile, it looks like there are conflicts to solve with master. Github reports 113 files changed, so I'm assuming something went wrong with the master merge. Or, more likely, this branch is quite old, and might have branched from master before jim did that force push to master a while ago, to fix some broken stuff.

If fixing conflicts becomes a burden, I suggest saving the commits from this work in a tmp branch, reset this branch to master, then cherry pick the commits from the tmp branch back.

**Comment by AaronDonahue** [comment]
@bartgol , this branch has been successfully rebased off of master and should hopefully pass the AT now.  The switch to ELM really through the rebase off, not an experience I would wish for many people :-)

Now the PR only wishes to change 12 files, which is more reasonable.

File: components/scream/src/physics/zm/CMakeLists.txt (hunk 3 above)
Comment (bartgol) [reply components/scream/src/physics/zm/CMakeLists.txt:7]: @mfossitt Should genf90 stuff be removed then? Or is it actually used?

**Review by bartgol — APPROVED** [review]
Since f90 is only temporary, I think there's no need to polish this PR any further. I only had one comment regarding genf90.

**Comment by AaronDonahue** [comment]
setting this back to WIP, something went wrong in the rebase which is why the tests are failing.  Meredith and I are working it out now.

**Comment by PeterCaldwell** [comment]
The mappy fail is clearly due to jenkins not being set up right still. Weaver isn't failing in a way that we've seen before (and the actual error message isn't included in the tail of the log) so I'm hesitant to ignore it. I just added AT:RETEST again. If it is spurious, it should pass this time.

**Comment by AaronDonahue** [comment]
For sure, I don't think we should merge till we sort out the AT failures.

**Comment by bartgol** [comment]
The mappy failure is due to the branch not being ahead of master. Unfortunately, the jenkins scripts are executed _before_ test-all-scream merges origin/master, so if your branch is old, it doesn't have the mappy scripts, and the AT craps out.

This is one of those scenario where a manual master merge is necessary, unfortunately.

**Comment by PeterCaldwell** [comment]
@AaronDonahue - it looks like the weaver fail is reproducible. Have you tested on Lassen recently? @bartgol - I think the build error message that's crashing this PR on weaver is happening earlier than the last 100 lines of the log. Could you find the actual error in the log and put it in a comment here?

**Comment by AaronDonahue** [comment]
@PeterCaldwell , I'll check on Lassen and report back.

**Comment by bartgol** [comment]
As for the failure on weaver, it appears some view in the inputs initializer is using the wrong memory space:

```
08:18:35 /home/e3sm-jenkins/weaver/workspace/SCREAM_PullRequest_Autotester_Weaver/440/scream/components/scream/src/physics/zm/zm_inputs_initializer.cpp(225): here
08:18:35 
08:18:35 /home/e3sm-jenkins/weaver/workspace/SCREAM_PullRequest_Autotester_Weaver/440/scream/externals/ekat/extern/kokkos/core/src/impl/Kokkos_ViewMapping.hpp(3271): error: static assertion failed with "View assignment must have compatible spaces"
```

**Comment by PeterCaldwell** [comment]
Thanks @bartgol . It sounds like a GPU-specific problem which should be reproducible on Lassen!

**Comment by AaronDonahue** [comment]
> As for the failure on weaver, it appears some view in the inputs initializer is using the wrong memory space:
> 
> ```
> 08:18:35 /home/e3sm-jenkins/weaver/workspace/SCREAM_PullRequest_Autotester_Weaver/440/scream/components/scream/src/physics/zm/zm_inputs_initializer.cpp(225): here
> 08:18:35 
> 08:18:35 /home/e3sm-jenkins/weaver/workspace/SCREAM_PullRequest_Autotester_Weaver/440/scream/externals/ekat/extern/kokkos/core/src/impl/Kokkos_ViewMapping.hpp(3271): error: static assertion failed with "View assignment must have compatible spaces"
> ```

Thanks for the heads up Luca.  The fix seems pretty simple.  We just pushed a bugfix that is hopefully correct:

```
-    Kokkos::View<Real*, Kokkos::LayoutRight, HostDevice> d_v = m_fields.at(zm_inputs[i]).get_view();
+    Kokkos::View<Real*, Kokkos::LayoutRight, DefaultDevice> d_v = m_fields.at(zm_inputs[i]).get_view();
```

**Comment by AaronDonahue** [comment]
Great, it looks like the latest round of AT passed for Blake and Weaver and the fail on Mappy was the expected one because the branch is not ahead of master.

**Comment by PeterCaldwell** [comment]
Should I merge this now? It looks like AT results are as expected. Are we sure this PR doesn't affect F-compset tests?

**Comment by AaronDonahue** [comment]
@PeterCaldwell , yes I think you can merge it.  All of the files touched are in `scream` so there shouldn't be any impact on F-Compsets.  Although, since there is no rush I'm happy to quickly run the scream_test_suite on the branch.
